docker/emptydrops/emptyDropsWrapper/emptyDropsWrapper.R

```diff
@@ -0,0 +1,211 @@
+    quit(save="no",status=1)
+}
+## Check dimensions
+if(any(dim(inputMatrix) == 0)) {
+    ## If the matrix is empty we can't run emptyDrops, write a empty table header
+    cat('Warning: one or more dimensions of the input matrix are empty. Generating empty result table.',file=stderr())
+    outputFile <- file(output_csv);
+    writeLines('"CellId","Total","LogProb","PValue","Limited","FDR","IsCell"',outputFile)
+    close(outputFile)
+    q(save="no",status=0)
+}
+## Check row and column names
+if(is.null(rownames(inputMatrix))) {
+    cat('Warning: rownames of the input matrix are empty',file=stderr())
+    quit(save="no",status=1)
+}
+if(is.null(colnames(inputMatrix))) {
+    cat('Warning: colnames of the input matrix are empty',file=stderr())
+    quit(save="no",status=1)
+}
+
+## Convert to CsparseMatrix if required
+## emptyDrops seems to work fine with dgRMatrix so this is not required for now
+## however it is here as it can help optimization in the future
+## if (class(inputMatrix) == 'dgRMatrix') {
+##   catv('Input Matrix is in dgRMatrix format. Converting...')
+##   inputMatrix <- as(inputMatrix, "CsparseMatrix")
+##   catv('done\n')
+## }
+
+## If requested transpose the input matrix
+## NOTE: dgRMatrix to dgCMatrix conversion and transposition
+## can be done in a single step by re-interpreting the indexes
+if (opt$transpose) {
+   catv('Transposing input matrix...')
+   inputMatrix <- Matrix::t(inputMatrix);
+   catv('done\n')
+}
+
+## Run emptyDrops with error handling
+catv('Running emptyDrops...')
+t0 <- Sys.time()
+tryCatch({
+    emptyDrops_result <- emptyDrops(m=inputMatrix,
+                                    lower=ed_param_lower,
+                                    niters=ed_param_niters,
+                                    test.ambient=ed_param_test.ambient,
+                                    ignore=ed_param_ignore,
+                                    alpha=ed_param_alpha,
+                                    BPPARAM=ed_param_BPPARAM)
+},error=function(e) {
+    cat('Error: an error occured while running emptyDrops!\n',file=stderr())
+    cat('Error: ', e$message,'\n',file=stderr())
+    quit(save="no",status=1)
+})
+t1 <- Sys.time()
+emptyDrop_runtime <- t1 - t0
+catv('done in', as.numeric(t1 - t0,units='secs'), 'seconds\n')
+
+catv('Preparing output table...')
+## Convert output from DataFrame to data.frame
+emptyDrops_result <- as.data.frame(emptyDrops_result)
+
+## Add the cell identifier as a column and remove from rownames
+emptyDrops_result$CellId <- rownames(emptyDrops_result)
+rownames(emptyDrops_result) <- NULL
+
+## Call the cells according to the cutoff
+emptyDrops_result$IsCell <- NA2FALSE(emptyDrops_result$FDR < FDRcutoff)
+
+## Define a column order
+colOrder <- c("CellId","Total", "LogProb", "PValue", "Limited", "FDR", "IsCell")
+emptyDrops_result <- emptyDrops_result[,colOrder]
```

Comment:
just to confirm, this is sorting the columns of the dataframe in the defined order right?

Reply:
no, this is just re-ordering the columns them selves, so that columnA is CellId, columnB Total etc. I don't reorder the rows to be sorted in any way, last I checked they are in the order of the input matrix and if any re-ordering is required it should be done when this is merged in the zarr

Reply:
Got it, thanks for the explanation!